core/src/main/java/org/infinispan/conflict/impl/DefaultConflictManager.java

```diff
@@ -0,0 +1,459 @@
+                  versionsMap.put(entry.getKey(), (InternalCacheValue<V>) rspVal);
+               } else {
+                  completableFuture.completeExceptionally(new CacheException(String.format("Unable to retrieve key %s from %s: %s", key, entry.getKey(), entry.getValue())));
+                  return;
+               }
+            }
+            completableFuture.complete(versionsMap);
+         });
+      }
+
+      @Override
+      public String toString() {
+         return "VersionRequest{" +
+               "key=" + key +
+               ", postpone=" + postpone +
+               '}';
+      }
+   }
+
+   private Predicate<? super Map<Address, InternalCacheEntry<K, V>>> filterConsistentEntries() {
+      return map -> map.values().stream().distinct().limit(2).count() > 1 || map.values().isEmpty();
+   }
+
+   private class ReplicaSpliterator implements Spliterator<Map<Address, InternalCacheEntry<K, V>>> {
+      private final LocalizedCacheTopology topology;
+      private int totalSegments, currentSegment;
+      private Iterator<Map<Address, InternalCacheEntry<K, V>>> iterator = Collections.emptyIterator();
+
+      ReplicaSpliterator(LocalizedCacheTopology topology) {
+         this(topology, 0, topology.getWriteConsistentHash().getNumSegments());
+      }
+
+      ReplicaSpliterator(LocalizedCacheTopology topology, int currentSegment, int totalSegments) {
+         this.topology = topology;
+         this.currentSegment = currentSegment;
+         this.totalSegments = totalSegments;
+      }
+
+      @Override
+      public boolean tryAdvance(Consumer<? super Map<Address, InternalCacheEntry<K, V>>> action) {
+         while (!iterator.hasNext()) {
+            if (currentSegment < totalSegments) {
+               try {
+                  if (trace)
+                     log.tracef("Attempting to receive all replicas for segment %s with topology %s", currentSegment, topology);
+                  List<Map<Address, InternalCacheEntry<K, V>>> segmentEntries = stateReceiver.getAllReplicasForSegment(currentSegment, topology).get();
+                  if (trace && !segmentEntries.isEmpty())
+                     log.tracef("Segment %s entries received: %s", currentSegment, segmentEntries);
+                  iterator = segmentEntries.iterator();
+                  currentSegment++;
+               } catch (InterruptedException e) {
+                  Thread.currentThread().interrupt();
+                  throw new CacheException(e);
+               } catch (ExecutionException | CancellationException e) {
+                  stateReceiver.stop();
+                  throw new CacheException(e.getMessage(), e.getCause());
+               }
+            } else {
+               return false;
+            }
+         }
+         action.accept(iterator.next());
+         return true;
+      }
+
+      @Override
+      public Spliterator<Map<Address, InternalCacheEntry<K, V>>> trySplit() {
+         int lo = currentSegment;
```

Comment:
Actually I'm not sure the original situation that you describe can occur with the fix in my comment, because the original spliterator always starts with its currentSegment set to 0 and the use of `~1` ensures that the starting segment is always even. The Spliterator docs state: `If the original thread hands a spliterator off to another thread for processing, it is best if that handoff occurs before any elements are consumed with tryAdvance()`, therefore assuming the java stream implementation follows this advice, the original situation you described should never occur as the `lo` variable will always be even when `trySplit()` is called.